cmd/branchingconfigmanagers/bugzilla-config-manager/main.go

```diff
@@ -0,0 +1,236 @@
+	agent := plugins.ConfigAgent{}
+	if err := agent.Load(configPath, []string{opts.pluginConfigDir}, "_pluginconfig.yaml", false); err != nil {
+		logrus.WithError(err).Fatal("failed to load Prow plugin configuration")
+	}
+
+	pluginConfig := agent.Config()
+	now := time.Now()
+	if opts.overwriteTime != nil {
+		now = *opts.overwriteTime
+	}
+	updatedBugzillaConfig, err := run(lifecycleConfig, pluginConfig.Bugzilla, now)
+	if err != nil {
+		logrus.WithError(err).Fatal("failed to reconcile Bugzilla config")
+	}
+
+	pluginConfig.Bugzilla = *updatedBugzillaConfig
+
+	pluginConfig, err = prowconfigsharding.WriteShardedPluginConfig(pluginConfig, afero.NewBasePathFs(afero.NewOsFs(), opts.pluginConfigDir))
+	if err != nil {
+		logrus.WithError(err).Fatal("failed to write plugin config shards")
+	}
+	pluginConfigRaw, err := yaml.Marshal(pluginConfig)
+	if err != nil {
+		logrus.WithError(err).Fatal("failed to serialize plugin config")
+	}
+
+	if err := ioutil.WriteFile(configPath, pluginConfigRaw, 0644); err != nil {
+		logrus.WithError(err).Fatal("failed to write main plugin config")
+	}
+}
+
+var releaseBranchPrefixes = []string{"openshift-", "release-"}
+var developmentDependentBugStates = &[]plugins.BugzillaBugState{{Status: "MODIFIED"}, {Status: "ON_QA"}, {Status: "VERIFIED"}}
+
+func run(lifecycleConfig ocplifecycle.Config, currentConfig plugins.Bugzilla, now time.Time) (*plugins.Bugzilla, error) {
+	developmentVersion, gaVersions, nonEOLNonGAVersions, err := extractFromConfig(lifecycleConfig, now)
+	if err != nil {
+		return nil, fmt.Errorf("failed to extract config from lifecycleConfig: %w", err)
+	}
+
+	return reconcileConfig(&currentConfig, developmentVersion, gaVersions, nonEOLNonGAVersions), nil
+}
+
+func extractFromConfig(lifecycleConfig ocplifecycle.Config, now time.Time) (developmentVersion *ocplifecycle.MajorMinor, gaVersions, nonEOLNonGAVersions []ocplifecycle.MajorMinor, err error) {
+	var errs []error
+	var developmentVersionFound bool
+
+	allNonEOLVersions := sets.String{}
+	gaVersionsSet := sets.String{}
+	for ocpProduct, productConfig := range lifecycleConfig {
+		if ocpProduct != "ocp" {
+			continue
+		}
+
+		for productVersion, events := range productConfig {
+			allNonEOLVersions.Insert(productVersion)
+			for _, event := range events {
+
+				// Future events, ignore
+				if event.When == nil || now.Before(event.When.Time) {
+					continue
+				}
+
+				// End of life version, nothing to do.
+				if event.Event == ocplifecycle.LifecycleEventEndOfLife {
+					allNonEOLVersions.Delete(productVersion)
+					break
+				}
+
+				if event.Event == ocplifecycle.LifecycleEventGenerallyAvailable {
+					gaVersionsSet.Insert(productVersion)
+					break
+				}
+
+				if event.Event == ocplifecycle.LifecycleEventOpen {
+					if developmentVersionFound {
+						errs = append(errs, fmt.Errorf("found multiple development versions: %s and %s", developmentVersion, productVersion))
+					}
```

Comment:
I think we'll falsely hit this error when we enforce the same criteria for versions in code freeze and the one that is open (like 4.9 and 4.10 are now). Timewise, the "older" release is not open, it just enforces the same criteria because last event that had any effect was "open".

Reply:
So much about being able to automatically do the correct thing by just falling through until a known condition is hit. Updated and added a testcase.